src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchWriter.cs

```diff
@@ -0,0 +1,569 @@
+            return this.RawOutputContext.FlushBuffersAsync()
+                .FollowOnSuccessWith(task => this.DisposeBatchWriterAndSetContentStreamRequestedState());
+        }
+#endif
+
+        /// <summary>
+        /// This method is called to notify that the content stream of a batch operation has been disposed.
+        /// </summary>
+        public override void BatchOperationContentStreamDisposed()
+        {
+            Debug.Assert(this.CurrentOperationMessage != null, "Expected non-null operation message!");
+
+            this.SetState(BatchWriterState.OperationStreamDisposed);
+            this.CurrentOperationRequestMessage = null;
+            this.CurrentOperationResponseMessage = null;
+            this.RawOutputContext.InitializeRawValueWriter();
+        }
+
+        /// <summary>
+        /// This method notifies the listener, that an in-stream error is to be written.
+        /// </summary>
+        /// <remarks>
+        /// This listener can choose to fail, if the currently written payload doesn't support in-stream error at this position.
+        /// If the listener returns, the writer should not allow any more writing, since the in-stream error is the last thing in the payload.
+        /// </remarks>
+        public override void OnInStreamError()
+        {
+            this.RawOutputContext.VerifyNotDisposed();
+            this.SetState(BatchWriterState.Error);
+            this.RawOutputContext.TextWriter.Flush();
+
+            // The OData protocol spec did not defined the behavior when an exception is encountered outside of a batch operation. The batch writer
+            // should not allow WriteError in this case. Note that WCF DS Server does serialize the error in XML format when it encounters one outside of a
+            // batch operation.
+            throw new ODataException(Strings.ODataBatchWriter_CannotWriteInStreamErrorForBatch);
+        }
+
+
+        /// <summary>
+        /// Starts a new changeset - implementation of the actual functionality.
+        /// </summary>
+        protected override void WriteStartChangesetImplementation()
+        {
+            // write pending message data (headers, response line) for a previously unclosed message/request
+            this.WritePendingMessageData(true);
+
+            // important to do this first since it will set up the change set boundary.
+            this.SetState(BatchWriterState.ChangesetStarted);
+            Debug.Assert(this.changeSetBoundary != null, "this.changeSetBoundary != null");
+
+            // reset the size of the current changeset and increase the size of the batch
+            this.ResetChangeSetSize();
+            this.InterceptException(this.IncreaseBatchSize);
+
+            // write the boundary string
+            ODataBatchWriterUtils.WriteStartBoundary(this.RawOutputContext.TextWriter, this.batchBoundary, !this.batchStartBoundaryWritten);

```

Comment:
>ODataBatchWriterUtils [](start = 12, length = 21)

I wonder if we should move ODataBatchWriterUtils to the multipartmixed namespace, and even rename to ODataMultipartMixedBatchWriterUtils, since these methods are specific to the Multipart/Mixed format.

Reply:
good idea! Moved & renamed.

---
In reply to: [141144087](https://github.com/OData/odata.net/pull/960#discussion_r141144087) [](ancestors = 141144087)